### ISSUE TITLE
Errors while load is interrupted

### ISSUE THREAD

**Issue by marcbaechinger** [google/shaka-player#1785]
Two questions at first to make sure I'm doing it right. :)

a) Am I required to cancel a pending load() in any other ways then calling load() again or calling unload()?
b) Am I required to wait for promises of load()/unload() to either being resolved or rejected before calling load()/unload() again?

After reading the docs I'm under the impression that I don't have to do this. Please let me know if I'm mistaken. Else the report below applies.

**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**

Yes, I found #1006 which is very useful. Also #1422.

**What version of Shaka Player are you using?**

2.5.0-beta2

**Can you reproduce the issue with our latest release version?**

Reproduced with 2.5.0-beta2

**Can you reproduce the issue with the latest code from `master`?**

Reproduced with 2.5.0-beta2

**Are you using the demo app or your own custom app?**

Custom app.

**If custom app, can you reproduce the issue using our demo app?**

No.

**What browser and OS are you using?**

Chrome on OSX.

**For embedded devices (smart TVs, etc.), what model and firmware version are you using?**

--

**What are the manifest and license server URIs?**

any dash/hls streams

**What did you do?**

I'm having a queue of items which are rapidly prepared one after the other in a circular, indefinite way with Shaka (or the plain video element). It's actually a stress test (like around 10 loads/unloads in a second).  I catch and ignore LOAD_INTERRUPTED errors which I expect to happen because, the load has been intentionally interrupted and Shaka needs to be prepared again. This accords to Shaka's definition of critical errors and which requires me to reprepare Shaka again.

**What did you expect to happen?**

I expect that all errors landing in the catch clause of Shaka are of type shaka.util.Error.

**What actually happened?**

In some cases on OSX I get a CRITICAL error which is not a Shaka error. I *think* it involves an HLS stream which seems to be pretty slow (Angel one) at least the initial time until the first frame rendered takes a while. Please note that after this error the video element is completely dead. calling shaka.load() or shaka.unload after this never resolves or rejects a promise anymore. Also setting the src attribute and calling functions on the video element directly does not do a thing anymore. 

I don't think that it's something which breaks shaka, but the video element. My aim would be to recognize this error in some ways so I can think about recovering by something like attaching a new media element (and maybe destroy and recreate shaka). That would be nice. But with the current AssertionError not being a shaka.util.Error this is difficult or leaves me in a situation in which I'm not sure how to best recover.

goog.asserts.AssertionError 
{
message: "Assertion failed: Wrong error type!", 
reportErrorToServer: true, 
messagePattern: "Assertion failed: Wrong error type!", 
stack: "AssertionError: Assertion failed: Wrong error type…ver-ks192jrn56912jsbhe12/app_desktop.js:43927:18)"}

This is caught in the catch clause of the load promise.

In one single case I got an OPERATION_ABORTED=7001 which should not be seen by apps:

shaka.util.Error
category: 7
code: 7001
data: []
handled: false
message: "Shaka Error PLAYER.OPERATION_ABORTED ()"
severity: 2
stack: "Error: Shaka Error PLAYER.OPERATION_ABORTED ()↵    at new shaka.util.Error (https://REMOVED-ks192jrn56912jsbhe12/app_desktop.js:27631:13)↵    at abort (https://REMOVED-ks192jrn56912jsbhe12/app_desktop.js:28571:25)↵    at shaka.util.AbortableOperation [as onAbort_] (https://REMOVED-ks192jrn56912jsbhe12/app_desktop.js:28619:15)↵    at shaka.util.AbortableOperation.abort (https://REMOVED-ks192jrn56912jsbhe12/app_desktop.js:28523:17)↵    at abort (https://REMOVED-ks192jrn56912jsbhe12/app_desktop.js:28575:19)↵    at shaka.util.AbortableOperation [as onAbort_] (https://REMOVED-ks192jrn56912jsbhe12/app_desktop.js:28619:15)↵    at shaka.util.AbortableOperation.abort (https://REMOVED-ks192jrn56912jsbhe12/app_desktop.js:28523:17)↵    at operations_.forEach (https://REMOVED-ks192jrn56912jsbhe12/app_desktop.js:29316:23)↵    at Array.forEach (<anonymous>)↵    at shaka.util.OperationManager.destroy (https://REMOVED-ks192jrn56912jsbhe12/app_desktop.js:29310:22)"
__proto__: Object

**Comment by marcbaechinger** [comment]
> In some cases on OSX I get a CRITICAL error which is not a Shaka error.

To be accurate. I get an error which is not a Shaka error and is error.severity !== RECOVERABLE because the AssertionError does not have the severity property.

**Comment by marcbaechinger** [comment]
Many thanks Aaron!

I did a lot of digging around this since. And I came to a similar conclusion that I have to call onload myself, even that I see that load does this for me (player.js:L782). 

However, I fear that I will see this AssertionError also when doing this, but I need to verify that first. 

Most beneficial for me would be that I can rest assured that Shaka is in a usable state after having caught an exception in load().throws() and that all these exception are of type shaka.util.Error - I currently get an AssertionError which puts Shaka in a state where I need to create a new Shaka instance with the same video element to recover. 

And yes, I can provide you with a link to a demo app which allows you to do kind of a stress test. To where should I send that url with instructions? Would be great to hear your opinion.

Thanks again!

**Comment by vaage** [comment]
You can send us a link privately via [shaka-player-issues@google.com](mailto:shaka-player-issues@google.com). To make it easier for us to connect the email with this issue, please ensure that the issue number is in the subject line.

You should not see the AssertionError out of `load` because `load` is the only function that will check the error and assert that it is `shaka.util.Error` and assert when its not. So really it is a question about why are we seeing a non-shaka error during the load process.

Once we can execute the stress test, I will drill into why we are seeing these two errors (OPERATION_ABORTED and AssertionError) and ensure they are resolved. The main area of my work these days is with the loading system, and I want to make sure it is as stable as possible, so ensuring you feel confident with what happens when a load fails, it a top priority of mine.

**Comment by vaage** [comment]
I am marking this as "waiting on response" while we wait to receive your email.

**Comment by marcbaechinger** [comment]
Many thanks Aaron! I sent the email. Let me know if I can assist you in any way.

**Comment by vaage** [comment]
Thanks for the email @marcbaechinger I have a better idea of what's happening now. It appears that when we destroy the HLS parser (and therefore cancel pending networking requests) the HLS parser thinks we have no partial HLS segment support.

Aside for the rapid `load` and `unload` usages, I think the errors found with them are all focused with [shaka.hls.HlsPlayer](https://github.com/google/shaka-player/blob/master/lib/hls/hls_parser.js).

What I suggest is that we open another issue to focus on the errors in [shaka.hls.HlsPlayer](https://github.com/google/shaka-player/blob/master/lib/hls/hls_parser.js), specially that it treats aborted networking request as the server failing segment requests. Having the separate issue will allow the conversation to be far more focused there and make it more discoverable by others.

@marcbaechinger, what are your thoughts on that?

**Comment by marcbaechinger** [comment]
Sound good! Thanks! 

You are right that when I've seen the AssertionError, there was always the HLS stream involved.

**Comment by vaage** [comment]
@marcbaechinger Are you okay with closing this issue since the core issues will be resolved with #1788 and #816?

**Comment by marcbaechinger** [comment]
Sure! That's fine! Thanks!
